feat: super-agent idempotent updates (NR-277039)

gsanchezgavier opened the pull request:
Supports running the recipe multiple times with the objective of updating the super-agent package version.
It uses the same approach of #1094 but leveraging [go-task status](https://taskfile.dev/reference/schema#task).

Also the PR removes some unsupported OS tests

gsanchezgavier commented:
this is how an update looks like:
``` shell
==> Installing New Relic Super Agent
A previous execution of this installation was detected. Some installation tasks will be skipped.
If you would like to run all the tasks, please remove the /etc/newrelic-super-agent/.nr-cli file and re-run the installation.
task: Task "cleanup" is up to date
task: Task "setup_infra_license" is up to date
task: Task "setup_infra_proxy" is up to date
Detecting available SSL ciphers...
 - SSLv3
 - TLSv1
 - TLSv1.2
 - TLSv1.3
Hit:1 http://ports.ubuntu.com/ubuntu-ports jammy InRelease
Hit:2 http://ports.ubuntu.com/ubuntu-ports jammy-updates InRelease
Hit:3 http://ports.ubuntu.com/ubuntu-ports jammy-backports InRelease
Hit:4 http://ports.ubuntu.com/ubuntu-ports jammy-security InRelease
Hit:5 https://download.newrelic.com/infrastructure_agent/linux/apt jammy InRelease
Hit:6 https://download.newrelic.com/preview/linux/apt jammy InRelease
Reading package lists...
W: https://download.newrelic.com/infrastructure_agent/linux/apt/dists/jammy/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key(8) for details.
W: https://download.newrelic.com/preview/linux/apt/dists/jammy/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key(8) for details.
Warning: apt-key is deprecated. Manage keyring files in trusted.gpg.d instead (see apt-key(8)).
OK
Hit:1 http://ports.ubuntu.com/ubuntu-ports jammy InRelease
Hit:2 http://ports.ubuntu.com/ubuntu-ports jammy-updates InRelease
Hit:3 http://ports.ubuntu.com/ubuntu-ports jammy-backports InRelease
Hit:4 https://download.newrelic.com/infrastructure_agent/linux/apt jammy InRelease
Hit:5 http://ports.ubuntu.com/ubuntu-ports jammy-security InRelease
Hit:6 https://download.newrelic.com/preview/linux/apt jammy InRelease
Reading package lists...
W: https://download.newrelic.com/infrastructure_agent/linux/apt/dists/jammy/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key(8) for details.
W: https://download.newrelic.com/preview/linux/apt/dists/jammy/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key(8) for details.
(Reading database ... 75856 files and directories currently installed.)
Preparing to unpack .../newrelic-super-agent_0.21.0_arm64.deb ...
Removed /etc/systemd/system/multi-user.target.wants/newrelic-super-agent.service.
Unpacking newrelic-super-agent (0.21.0) over (0.20.0) ...
Setting up newrelic-super-agent (0.21.0) ...
Created symlink /etc/systemd/system/multi-user.target.wants/newrelic-super-agent.service → /lib/systemd/system/newrelic-super-agent.service.
task: Task "update_otel_license_key" is up to date
task: Task "config_supervisors" is up to date
task: Task "config_fleet_id" is up to date
task: Task "config_opamp" is up to date
task: Task "config_super_agent_auth" is up to date
task: Task "config_host_monitoring" is up to date
task: Task "update_otel_mem_limit" is up to date
task: Task "update_otel_end_point" is up to date
task: Task "migrate_newrelic_infra_config" is up to date
✔ Installing New Relic Super Agent
   Installed 

```

kang-makes reviewed (commented):
I left some comments for you while you work on making those test pass :D

gsanchezgavier commented on recipes/newrelic/infrastructure/super-agent/debian.yml:
```diff
@@ -582,9 +603,19 @@ install:
           if [ $IS_AGENT_INSTALLED -eq 0 ] ; then
             echo "The newrelic super agent has not started after installing. Please try again later, or see our documentation for installing manually https://docs.newrelic.com/docs/using-new-relic/cross-product-functions/install-configure/install-new-relic" >&2
+
+            if [ {{.IS_SYSTEMCTL}} -gt 0 ]; then
```
I moved the vars to the root level so they can be accesible from any task to avoid code duplication. 
regarding removing that i would leave it out of scope for this PR.

kang-makes replied:
Fair enough.

gsanchezgavier commented on recipes/newrelic/infrastructure/super-agent/debian.yml:
```diff
@@ -183,13 +197,16 @@ install:
 
     cleanup:
+      # skipping this task if .nr-cli exists
+      status: ["test -f /etc/newrelic-super-agent/.nr-cli"] 
```
fair enough , was looking nice to me on the IDE but i guess isn't the case here

gsanchezgavier commented on recipes/newrelic/infrastructure/super-agent/debian.yml:
```diff
@@ -350,6 +367,7 @@ install:
 
     # If configured to do so, migrate the newrelic-infra configuration for usage with the super agent
     migrate_newrelic_infra_config:
+      status: ["test -f /etc/newrelic-super-agent/.nr-cli"]
```
yes we do not want to migrate configs after a succesfull install , I would prefer to keep the same approach even if there is another flag under checking for other cases.

gsanchezgavier commented on recipes/newrelic/infrastructure/super-agent/debian.yml:
```diff
@@ -359,12 +377,14 @@ install:
 
     # Add NR Ingest Key as ENV Var for the Super Agent Systemd service (used by Infra Agent and Otel)
     update_otel_license_key:
+      status: ["test -f /etc/newrelic-super-agent/.nr-cli"] 
```
If i understood correctly , you would like to support not only upgrade of the package versions but also a key rotation. Agree that would be a nice case to discuss with @josemore , also I would try to increase automatic testing of different scenarios before increasing the number of features we have here. 
For now let's keep it out of the scope for this PR

kang-makes reviewed (dismissed):
LGTM

Avinash-Sanpala reviewed (commented):
@gsanchezgavier  Changes look good to me. I will perform a manual test on Monday and will merge it as currently the super agent tests are failing because of the organisation id is not passed as an argument to the cli command in CI and this implementation is currently in progress.

gsanchezgavier commented:
@Avinash-Sanpala  i just rebased from main, which comes with the org_id and I see the tests passing, except for:
- removed eol OS  which we don't want to test anymore (expected 👍 )
- Ubuntu16 which has an incompatible python version, for that version of Ansible
Could you have a look at this issue with Python ?
Thanks

Avinash-Sanpala commented:
> @Avinash-Sanpala i just rebased from main, which comes with the org_id and I see the tests passing, except for:
> 
> * removed eol OS  which we don't want to test anymore (expected 👍 )
> * Ubuntu16 which has an incompatible python version, for that version of Ansible
>   Could you have a look at this issue with Python ?
>   Thanks

@gsanchezgavier Yes this org_id issue is fixed now and most of the tests are expected to pass except few that could fail with know issues. Sure will have a look. Thanks!

Avinash-Sanpala reviewed (approved):
LGTM and could see all the supporting test cases are running successfully
